src/src/com/microsoft/aad/adal/IdToken.java

```diff
@@ -23,28 +23,145 @@
 
 package com.microsoft.aad.adal;
 
+import java.io.UnsupportedEncodingException;
+import java.util.HashMap;
+import java.util.Iterator;
+
+import org.json.JSONException;
+import org.json.JSONObject;
+
+import android.util.Base64;
+
 class IdToken {
-    String mSubject;
 
-    String mTenantId;
+    final private static String TAG = "IdToken";
+
+    private String mSubject;
+
+    private String mTenantId;
+
+    private String mUpn;
+
+    private String mGivenName;
+
+    private String mFamilyName;
+
+    private String mEmail;
+
+    private String mIdentityProvider;
+
+    private String mObjectId;
+
+    private long mPasswordExpiration;
+
+    private String mPasswordChangeUrl;
+
+    public IdToken(String idtoken) throws AuthenticationException {
+        // Message segments: Header.Body.Signature
+        HashMap<String, String> responseItems = this.parseJWT(idtoken);
+
+        if (responseItems != null && !responseItems.isEmpty()) {
+            this.mSubject = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_SUBJECT);
+            this.mTenantId = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_TENANTID);
+            this.mUpn = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_UPN);
+            this.mEmail = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_EMAIL);
+            this.mGivenName = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_GIVEN_NAME);
+            this.mFamilyName = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_FAMILY_NAME);
+            this.mIdentityProvider = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_IDENTITY_PROVIDER);
+            this.mObjectId = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_OBJECT_ID);
+            String expiration = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_PASSWORD_EXPIRATION);
+
+            if (!StringExtensions.IsNullOrBlank(expiration)) {
+                this.mPasswordExpiration = Long.parseLong(expiration);
+            }
+
+            this.mPasswordChangeUrl = responseItems.get(AuthenticationConstants.OAuth2.ID_TOKEN_PASSWORD_CHANGE_URL);
+        }
+    }
+
+    public String getSubject() {
+        return mSubject;
+    }
+
+    public String getTenantId() {
+        return mTenantId;
+    }
+
+    public String getUpn() {
+        return mUpn;
+    }
 
-    String mUpn;
+    public String getGivenName() {
+        return mGivenName;
+    }
+
+    public String getFamilyName() {
+        return mFamilyName;
+    }
 
-    String mGivenName;
+    public String getEmail() {
+        return mEmail;
+    }
+
+    public String getIdentityProvider() {
+        return mIdentityProvider;
+    }
+
+    public String getObjectId() {
```

Comment:
ObjectId comes directly from the server and is a core AAD concept.  I don't want to re-document AAD concepts doced elsewhere in this code.